### PR TITLE
[KM157] Delete ArgoCD and all resources associated with the service

### PR DESCRIPTION
**Pull request by paulbes**
**NB:** This PR needs to be merged before #302 

Delete ArgoCD and all resources associated with the service

## Description
Clean up all ArgoCD resources by deleting the `argocd` namespace, removing the certificate and any dangling DNS entries, etc. Try to leave the AWS cluster of the user in as clean a state as possible.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the release notes (for the next release).

### CODE REVIEW

File: pkg/client/core/service_argocd_client.go [hunk 1]
```diff
@@ -28,6 +28,53 @@ type argoCDService struct {
...
+
+	defer func() {
+		err = s.spinner.Stop()
+	}()
+
+	info := s.state.GetArgoCD(opts.ID)
+
+	err = s.manifest.DeleteNamespace(ctx, api.DeleteNamespaceOpts{
+		ID:        opts.ID,
+		Namespace: "argocd",
```
Comment (deifyed) [inline pkg/client/core/service_argocd_client.go:45]: Should probably be something like config.DefaultArgoCDNamespace

File: pkg/client/core/service_argocd_client.go [hunk 2]
```diff
@@ -28,6 +28,53 @@ type argoCDService struct {
...
+	})
+	if err != nil {
+		return err
+	}
+
+	err = s.cert.DeleteCognitoCertificate(ctx, api.DeleteCognitoCertificateOpts{
```
Comment (deifyed) [inline pkg/client/core/service_argocd_client.go:51]: Should we delete Cognitos certificate here? We'll probably use Cognito for other things as well down the line

File: pkg/client/manifest_client.go [hunk 3]
```diff
@@ -22,11 +22,16 @@ type CreateExternalSecretOpts struct {
 // ManifestService implements the business logic
 type ManifestService interface {
 	CreateExternalSecret(ctx context.Context, opts CreateExternalSecretOpts) (*ExternalSecret, error)
+	// I don't like this, but I think ManifestService is an equally big problem
```
Comment (deifyed) [inline pkg/client/manifest_client.go:25]: Why don't you like it?

File: pkg/kube/kube.go [hunk 4]
```diff
@@ -32,11 +44,11 @@ type Kuber interface {
 // Kube contains state for communicating with
 // a kubernetes cluster
 type Kube struct {
-	KubeConfigPath string
-	ClientSet      *kubernetes.Clientset
-	RestConfig     *rest.Config
-	Ctx            context.Context
-	Log            *logrus.Logger
+	Provider   ClientSetProvider
```
Comment (deifyed) [inline pkg/kube/kube.go:47]: Whats a Kubernetes client set?

File: pkg/api/core/cloudprovider/aws/certificate_aws.go [hunk 5]
```diff
@@ -14,7 +14,7 @@ type certificate struct {
 	provider v1alpha1.CloudProvider
 }
 
-func (c *certificate) DeleteCertificate(opts api.DeleteCertificateOpts) error {
+func (c *certificate) DeleteCognitoCertificate(opts api.DeleteCognitoCertificateOpts) error {
```
Comment (deifyed) [inline pkg/api/core/cloudprovider/aws/certificate_aws.go:17]: Why does Cognito require its own unique certificate deletion functionality?

File: pkg/api/core/handler_api.go [hunk 6]
```diff
@@ -393,6 +414,9 @@ func InstrumentEndpoints(logger *logrus.Logger) EndpointOption {
 			CreateAWSLoadBalancerControllerPolicy:         logmd.Logging(logger, strings.Join([]string{managedPoliciesTag, awsLoadBalancerControllerTag}, "/"), "create")(endpoints.CreateAWSLoadBalancerControllerPolicy),
 			DeleteAWSLoadBalancerControllerPolicy:         logmd.Logging(logger, strings.Join([]string{managedPoliciesTag, awsLoadBalancerControllerTag}, "/"), "delete")(endpoints.DeleteAWSLoadBalancerControllerPolicy),
 			CreateAWSLoadBalancerControllerHelmChart:      logmd.Logging(logger, strings.Join([]string{helmTag, awsLoadBalancerControllerTag}, "/"), "create")(endpoints.CreateAWSLoadBalancerControllerHelmChart),
+			DeleteCertificate:                             logmd.Logging(logger, certificateTag, "delete")(endpoints.DeleteCertificate),
+			DeleteNamespace:                               logmd.Logging(logger, strings.Join([]string{kubeTag, namespaceTag}, "/"), "delete")(endpoints.DeleteNamespace),
+			DeleteCognitoCertificate:                      logmd.Logging(logger, strings.Join([]string{certificateTag, cognitoTag}, "/"), "delete")(endpoints.DeleteCognitoCertificate),
```
Comment (ivaruf) [inline pkg/api/core/handler_api.go:419]: ... what is wrong with the old "someting"+"/"+"else"?  ... well i guess I understand what this does ;)

**Review by ivaruf — COMMENTED** [review]
Probably should get rid of
"pkg/cognito/authcertdeleter.go" as part of this PR

File: pkg/client/manifest_client.go (hunk 3 above)
Comment (yngvark) [inline pkg/client/manifest_client.go:25]: Can you specify what "this" means, so the next developer has a better idea what the problem is?

**Review by yngvark — APPROVED** [review]

File: pkg/client/core/service_argocd_client.go (hunk 2 above)
Comment (paulbes) [reply pkg/client/core/service_argocd_client.go:51]: Oh shit, nice catch. Absolutely not.

File: pkg/client/manifest_client.go (hunk 3 above)
Comment (paulbes) [reply pkg/client/manifest_client.go:25]: The name `Manifest` was initially derived from the Kubernetes manifest lingo. I am not certain this is conveyed, maybe mirroring something like `Kube` or similar would be better? I think there are things around kube/manifest that could use a proper review

File: pkg/kube/kube.go (hunk 4 above)
Comment (paulbes) [reply pkg/kube/kube.go:47]: Something that knows how to communicate with the Kubernetes API server, generally it also refers to a specific client

File: pkg/api/core/cloudprovider/aws/certificate_aws.go (hunk 5 above)
Comment (paulbes) [reply pkg/api/core/cloudprovider/aws/certificate_aws.go:17]: I am not sure, this functionality existed prior to my refactor, I just moved it. My assumption was that the auth cert had to be disassociated before cognito could be deleted

File: pkg/client/manifest_client.go (hunk 3 above)
Comment (paulbes) [reply pkg/client/manifest_client.go:25]: Good point, done!

**Comment by paulbes** [comment]
> Probably should get rid of
"pkg/cognito/authcertdeleter.go" as part of this PR

Why isn't it required @ivaruf?

File: pkg/client/manifest_client.go (hunk 3 above)
Comment (deifyed) [reply pkg/client/manifest_client.go]: Agreed. Manifest confuses me. A KubeService/KubernetesServer would make alot more senses (at least to me)

**Pull request merged**